dataflow/src/org/checkerframework/dataflow/cfg/DOTCFGVisualizer.java

```diff
@@ -65,25 +70,97 @@ public void init(Map<String, Object> args) {
+
+        String dotgraph = generateDotGraph(cfg, entry, analysis);
+
+        String dotfilename = dotOutputFileName(cfg.underlyingAST);
+        // System.err.println("Output to DOT file: " + dotfilename);
+
+        try {
+            FileWriter fstream = new FileWriter(dotfilename);
+            BufferedWriter out = new BufferedWriter(fstream);
+            out.write(dotgraph);
+            out.close();
+        } catch (IOException e) {
+            ErrorReporter.errorAbort("Error creating dot file: " + dotfilename +
+                    "; ensure the path is valid", e);
+        }
+
+        Map<String, Object> res = new HashMap<>();
+        res.put("dotFileName", dotfilename);
+
+        return res;
+    }
+
+    /**
+     * generate the dot string write to dot file
+     * @param cfg
+     * @param entry
+     * @param analysis
+     * @return
+     */
+    protected String generateDotGraph(ControlFlowGraph cfg, Block entry,
+        /*@Nullable*/ Analysis<A, S, T> analysis) {
+        StringBuilder sbDigraph = new StringBuilder();
+        StringBuilder sbNode = new StringBuilder();
+        StringBuilder sbEdge = new StringBuilder();
         Set<Block> visited = new HashSet<>();
-        Queue<Block> worklist = new LinkedList<>();
-        Block cur = entry;
-        visited.add(entry);
 
         // header
-        sb1.append("digraph {\n");
-        sb1.append("    node [shape=rectangle];\n\n");
+        sbDigraph.append("digraph {\n");
+
+        /*generateDotEdges must called BEFORE generateDotNodes,
+         * since DoteNodes needs the information in 'Set<Block> visited' that created in generateDotEdges()  */
+        generateDotEdges(sbEdge, visited, entry);
+
+        generateDotNodes(sbNode, visited, cfg, analysis);
 
+        sbDigraph.append(sbNode);
+        sbDigraph.append(sbEdge);
+
+        // footer
+        sbDigraph.append("}\n");
+
+        return sbDigraph.toString();
+    }
+
+    private void generateDotNodes(StringBuilder sbNode, Set<Block> visited,
+			ControlFlowGraph cfg, Analysis<A, S, T> analysis) {
+        IdentityHashMap<Block, List<Integer>> processOrder = getProcessOrder(cfg);
+
+        sbNode.append("    node [shape=rectangle];\n\n");
+     // definition of all nodes including their labels
+        for (Block v : visited) {
+            sbNode.append("    " + v.getId() + " [");
+            if (v.getType() == BlockType.CONDITIONAL_BLOCK) {
+                sbNode.append("shape=polygon sides=8 ");
+            } else if (v.getType() == BlockType.SPECIAL_BLOCK) {
+                sbNode.append("shape=oval ");
+            }
+            sbNode.append("label=\"");
+            if (verbose) {
+                sbNode.append("Process order: " + processOrder.get(v).toString().replaceAll("[\\[\\]]", "") + "\\n");
+            }
+            sbNode.append(visualizeBlock(v, analysis).replace("\\n", "\\l")
+                    + " \",];\n");
+        }
+
+        sbNode.append("\n");
```

Comment:
Consistently use four spaces for indentation. Something is off here and in the next method.